runtime/vm/resolvesupport.cpp

```diff
@@ -1839,6 +1839,102 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	j9object_t value = ramCPEntry->value;
+
+	J9Class *ramClass = ramCP->ramClass;
+	J9ROMClass *romClass = ramClass->romClass;
+	J9ROMConstantDynamicRef *romConstantRef = (J9ROMConstantDynamicRef*)(J9_ROM_CP_FROM_CP(ramCP) + cpIndex);
+	J9SRP *callSiteData = (J9SRP *) J9ROMCLASS_CALLSITEDATA(romClass);
+	U_16 *bsmIndices = (U_16 *) (callSiteData + romClass->callSiteCount);
+	U_16 *bsmData = bsmIndices + romClass->callSiteCount;
+	U_32 bsmIndex = romConstantRef->bsmIndexAndCpType >> J9DescriptionCpTypeShift;
+	J9ROMNameAndSignature* nameAndSig = SRP_PTR_GET(&romConstantRef->nameAndSignature, J9ROMNameAndSignature*);
+
+	J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
+	J9Class *throwable = internalFindKnownClass(vmThread, J9VMCONSTANTPOOL_JAVALANGTHROWABLE, J9_FINDKNOWNCLASS_FLAG_INITIALIZE);
+	UDATA resolved = 0;
+
+retry:
+	/* Check if already resolved */
+	if (NULL != value) {
+		resolved = 1;
+	} else if (NULL != ramCPEntry->exception) {
+		if (isSameOrSuperClassOf(throwable, J9VM_J9CLASS_FROM_HEAPCLASS(vmThread, ramCPEntry->exception))) {
+			/* Resolved with exception */
+			VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+			resolved = 1;
+		} else if (ramCPEntry->exception == vmThread->javaVM->voidReflectClass->classObject) {
+			/* Resolved with null reference */
+			resolved = 1;
+		} else if (ramCPEntry->exception != vmThread->threadObject) {
+			/* Another thread is currently trying to resolve the entry */
+			// yield(vmThread);
+			goto retry;
+		}
+	}
+
+	/* Enter if not previously resolved */
+	if (0 == resolved) {
+		/* Set exception to current threadObject as a mutex to prevent other threads from entering */
+		if ((ramCPEntry->exception != vmThread->threadObject) &&
+			(0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, ramClass, &ramCPEntry->exception, NULL, vmThread->threadObject))
+		) {
+			goto retry;
+		}
+
+		/* Walk bsmData */
+		for (U_32 i = 0; i < bsmIndex; i++) {
+			bsmData += bsmData[1] + 2;
+		}
+
+		/* Invoke BootStrap method*/
+		sendResolveConstantDynamic(vmThread, ramCP, cpIndex, nameAndSig, bsmData);
+		value = (j9object_t)vmThread->returnValue;
+
```

Comment:
better to move the `(ramCPEntry->exception != vmThread->threadObject)` check here, no point even running the rest of the code if someone else has resolved it

probably need read barrier before the check

Reply:
fixed